src/VaultSharp/V1/SecretsEngines/PKI/PKISecretsEngineProvider.cs

```diff
@@ -126,5 +128,54 @@ public async Task<Secret<CertificateData>> ReadDefaultIssuerCertificateChainAsyn
             var certificateDataSecret = await _polymath.MakeVaultApiRequest<Secret<CertificateData>>(pkiBackendMountPoint ?? _polymath.VaultClientSettings.SecretsEngineMountPoints.PKI, path, HttpMethod.Get, unauthenticated: true).ConfigureAwait(_polymath.VaultClientSettings.ContinueAsyncTasksOnCapturedContext);
             return certificateDataSecret;
         }
+
+        public async Task<Secret<GenerateRootResponse>> GenerateRootAsync(string type, GenerateRootRequest generateRootRequest, string pkiBackendMountPoint = null)
+        {
+            var path = $"/root/generate/{type}";
+            var generatedRootDataSecret = await _polymath.MakeVaultApiRequest<Secret<GenerateRootResponse>>(pkiBackendMountPoint ?? _polymath.VaultClientSettings.SecretsEngineMountPoints.PKI, path, HttpMethod.Post, generateRootRequest).ConfigureAwait(_polymath.VaultClientSettings.ContinueAsyncTasksOnCapturedContext);
+            return generatedRootDataSecret;
+        }
+
+        public async Task<Secret<PKIRoleNames>> ListRolesAsync(string pkiBackendMountPoint = null)
+        {
+            var path = "/roles?list=true";
+            var roleNamesSecret = new Secret<PKIRoleNames>() { Data = new PKIRoleNames() { Keys = new List<string>() } };
+            
+            try
+            {
+                roleNamesSecret = await _polymath.MakeVaultApiRequest<Secret<PKIRoleNames>>(pkiBackendMountPoint ?? _polymath.VaultClientSettings.SecretsEngineMountPoints.PKI, path, HttpMethod.Get).ConfigureAwait(_polymath.VaultClientSettings.ContinueAsyncTasksOnCapturedContext);                
+            }
+            catch (VaultApiException ex)
+            {
+                // if the role list is empty, we will get a 404 response with { "errors": [] }
+                // make sure it's that error only, and return the empty list if so
+                if (ex.StatusCode != 404)
+                {
+                    throw;
+                }                
```

Comment:
I am not sure about this. A HTTP Status Code 404 can be returned even in the case when the `mountPoint` is incorrect. Handling the 404 exception might be an application layer decision. I will let @rajanadar  comment.